### ISSUE TITLE
Training + Inference to predict counts

### ISSUE THREAD

**Issue by faroit** [gokceneraslan/neuralnet_countmodels#1]
First I would like to say that your notebook is fantastic, this helped me a lot and I was able to run this immediately. I have two major questions that may both be related. Just a warning, as an engineer I find it difficult to understand the mathematical background of the count models, so I hope you can help me clarify a few things here ;-)

## 1. Inference

Lets start with the very simple example of poisson regression in machine learning context. Lets say I want to learn the number of objects in an image. To actually train a model, I will set the output to `Dense(1)` and set the activation function to `K.exp`. I'd then use the poisson loss to estimate the `lambda`  parameter. That way my output is comparing the true count `y_true`, which is an integer, to the predicted `_lambda`.  For inference to get the real count value I compute the mode from the poisson probability mass distribution (depending on how the pmf is defined it might need to be shifted):

```python
y_pred = np.argmax(poisson.pmf(np.arange(max_count), _lambda))
```

So first question do I get this right? 

## 2. Negative Binomial

related to how I would do inference using the poisson regression model, I wonder why this is different for the negative binomial:

* why is `theta` a trainable network parameter instead of an additional output? Wouldn't that assume that, once trained, the dispersion is constant for the whole dataset? 
* If that is the case, how would I do the inference then?

This brings me to this paper: [DeepSetNet: Predicting Sets with Deep Neural Networks](https://arxiv.org/pdf/1611.08998.pdf). They are proposing to use negative binomial loss as well. However I have a hard time understanding how this actually works in a complete system.

> To learn the cardinality distribution, we use the same
VGG-16 network as above and modify the final fully connected
layer to predict 2 values followed by two weighted
sigmoid activation functions for α and β. It is important
to note, that the outputs must be positive to describe a
valid Gamma distribution. We therefore also append two
weighted sigmoid transfer functions with weights αM, βM
to ensure that the values predicted for α and β are in a
valid range. Our model is not sensitive to these parameters
and we set their values to be large enough (αM = 160
and βM = 20) to guarantee that the mode of the distribution
can accommodate the largest cardinality existing in the
dataset.

What is your opinion on this? Would would be the simplest way to use NB for counting?

**Comment by gokceneraslan** [comment]
Thanks for the nice thoughts, I'm actually surprised that you found this notebook :)

1. I'm not sure that I understand the code, what is `max_count`? The output of the neural net is `y_pred = exp(WX+b)` and this already represents our estimate of the `lambda`, the mean of Poisson. So, there is no other additional _lambda. Poisson loss then calculates the likelihood of this parameter (y_pred) given the data (y_true). That's exactly same as using mean squared error, which is derived from the negative log likelihood of Normal pdf. So when we use mse,  we estimate the mean parameter of Normal distribution (output of the model, y_pred) that is most likely to generate our data (y_true). Using Poisson loss in a neural net is actually a poisson regression, so you can read any resource about it to clarify the concept. We can replace linear part of poisson regression with MLPs. By the way, you do not need the mode at all, if you do you can use y_pred, the mean, to calculate the mode of poisson.

2. That's a good point. In linear negative binomial regression, dispersion parameter is typically not conditioned on data (see `glm.nb` [function](https://stat.ethz.ch/R-manual/R-devel/library/MASS/html/glm.nb.html) in R), but in case of neural models, I believe that it makes sense to make it a function of data as well. I actually implemented it [here](https://github.com/gokceneraslan/autoencoder/blob/master/autoencoder/network.py#L146). It's just like the `Pi` layer, an output of the model.

Feel free to ask if you have other questions.

**Comment by faroit** [comment]
thanks for your reply! Some follow up questions:

> I'm not sure that I understand the code, what is max_count? 

if the poisson distribution is defined as:
![](https://wikimedia.org/api/rest_v1/media/math/render/svg/c22cb4461e100a6db5f815de1f44b1747f160048)

`max_count` would be that particular `k` that I would test for to get the most probable count.

> The output of the neural net is y_pred = exp(WX+b) and this already represents our estimate of the lambda, the mean of Poisson. 

okay, since my target is actually the integer count, I would name the output of my NN `_lambda` instead of `y_pred` and use the inference 'lookup' given as mentioned above to generate integer counts `y_pred`. This differs from your example where the target is predicting the mean. 

So using your definition, I would use 

```
y_count = np.argmax(poisson.pmf(np.arange(max_count), y_pred))
```

And just to clarify: in your case, `y_true` is a real or integer?

> By the way, you do not need the mode at all, if you do you can use y_pred, the mean, to calculate the mode of poisson.

I don't think I fully understand what you mean by this. Is this because we were talking about different targets?

>  but in case of neural models, I believe that it makes sense to make it a function of data as well. I actually implemented it here. It's just like the Pi layer, an output of the model.

Oh thats great. Regarding your implementation: 

* So `pi` is the parameter related to the Zero inflated model, so since I am only interested in NB, I would only need to feed theta from the `disp` output layer into the loss, right?
* Your dispersion output activation is `1.0/(K.exp(x)+1e-10`, I thought this would also work with `exp` or even `sigmoid`, am I wrong?
* What is the slicing layer needed for? Lets say I would only want to estimate the NB parameters, can't I just use naïvely use `dense(2)` and use `exp` or a scaled `sigmoid` for the output then?

**Comment by gokceneraslan** [comment]
1. If I understand correctly, you are trying to make integer predictions basically _after_ the training, right? Since the network output (`_lambda = exp(WX+b)`) is a continuous non-negative value representing the mean of Poisson, you can just use the mode of the poisson with mean of `_lambda` via `floor(_lambda)`. I think that's same as what you do with `np.argmax`. Regarding your question, `y_true` values are the integer labels of the data, which are samples from Poisson. I wouldn't use poisson/nb if they were continuous.

2.   
    - You're right, you need the output of `disp` layer for the dispersion.
    - It would work with exp as well, yes. It's just a different nomenclature. Some people use `var=mean + mean^2 * disp`, whereas some use `var=mean + mean^2 / disp` formula for the variance of NB. Former corresponds to poisson when disp=0, whereas latter is possion if `disp=\infty`. `sigmoid` activation on the other hand is a bad idea because the dispersion can (and in many cases will) be bigger than 1.
    - Slicing layer is just a hack I use to insert other trainable parameters to Keras. Yes, you can also use dense(2). Then you need to handle it in the loss function, which will split `y_pred` and treat the first value as nb mean and the second as dispersion. It might get messy when you have multiple ( as in multitask/multilabel/multivariate) NB outputs. Then, you need a closure or class in order not to hard-code the output dimension in the loss function.

**Comment by faroit** [comment]
Okay thanks again. I think the inference is clear now. And yes the mode can easily predicted using the floor function instead using the `pmf`. Now, when I train the poisson regression on very simple toy data ([see code](https://gist.github.com/faroit/a04932f5f25159e500f6303af3483145)), I still think there is something wrong. The input is `(nb_samples, 10)` where each sample has a number of elements equal to `1.0`. The task is to count these elements. From the results you can see that `lambda` is a good estimate but the floor function ruins the integer count prediction. A simple rounding would work better here. I know that the example is not a good fit, because the samples are not poisson distributed. But still, shouldn't the results be better than using `mse` loss? Any idea?

![poisson_inference](https://cloud.githubusercontent.com/assets/72940/25660156/418458f4-300b-11e7-80ed-7adf803b0507.png)

One more question, I might want to include some of your implementation details you are doing in the autoencoder project for the results of a paper, do you have any paper that I could cite? otherwise I would put you in the acknowledgements... ;-)

**Comment by faroit** [comment]
@gokceneraslan okay, just out of curiosity I've implemented both (P/NB) in a larger framework with real data. There the poisson loss gives slightly better results compared to linear/rms loss as it is indicated in the `deepsetnet` paper. In any case I would be really grateful if you could have a quick look at the code mentioned above. I still have the feeling there is something suspiciously wrong...

**Comment by tblazina** [comment]
@gokceneraslan, Thanks a lot for putting this notebook together, I'm trying to predict count data which has a lot of zero values and I was surprised how this really isn't discussed regularly when introducing people to doing regression with NNs (or in general with doing GLMs). Not sure if you are still responding to questions/comments on this notebook but... I had a question or two. In the notebook you add the eps variable to a lot of numbers throughout the note book but it isn't discussed (you have a comment in the NB class '# for numerical stability' when you define it), could you elaborate on why you are adding this factor in so many places? I had also had a question, maybe it's something basic i'm missing but when you say want to predict the output of a network as an integer then you would take the mode of the poisson distribution with a mean value (_lambda), which is floor(lambda) (https://math.stackexchange.com/questions/246496/the-mode-of-the-poisson-distribution), correct?  So my question then would be, how do you convert the  output of the network to an integer when using the negative binomial (or the zero inflated model). I see calculating the mode is fairly straight forward (https://en.wikipedia.org/wiki/Negative_binomial_distribution) but since you show the alternative parameterization with mean and dispersion, I get a bit lost.  Any input would be greatly appreciated and thanks again for the great notebook!

**Comment by gokceneraslan** [comment]
Hey, @tblazina, epsilon is added just to prevent evaluation of log(0.0) or lgamma(0.0) which leads to infinite log likelihood. Mode of NB can be used for converting the output to integers, yes. Here is how I calculate it using the alternative parameterization:

```python
mode = np.floor(mu*((theta-1)/theta)).astype(np.int)
mode[mode < 0] = 0
```
where mu and theta correspond to the mean and dispersion.

**Comment by tblazina** [comment]
thanks @gokceneraslan for the quick response. I think this makes sense. I have another question if you don't mind? In your multivariate NB regression example, you use tensor flow slim to estimate the weight, biases and theta (the dispersion of NB), but i guess i'm not understanding then how you are estimating mu? Am I correct in that this is simply tf.exp(W * X + b)?  If so then I'm a bit confused, so it would return 1000 x 10 matrix, and each row would contain 10 estimates of the mean of a NB distribution which we are modeling?

Secondly, if we wanted to create the output as counts then you would do as you said above, but theta is only estimated once for (so it is a 1 x10 matrix) for the whole model. I think this is what @faroit was discussing with you above when he asked about treating the dispersion also as an output for the model, which in case you would also have a 1000 x 10 matrix and then each estimate of the would have a mean and a dispersion and then you could have an output calculated as a count and an associated uncertainty for each measurement from the dispersion, am I correct in my summary?

If so i'll dig into your auto encoder code you mention above. The notion of having a point estimate and an uncertainty at each point is super cool!

or maybe something still isn't clicking for me :-P

Ultimately i'm trying to implement: https://arxiv.org/pdf/1704.04110.pdf, doing forecasting on count data,  so like the idea would be to have a bunch of time series of count data and other covariates (day of year, categorical data ect.) that are fed into an RNN network with LSTM cells that would predict sequences which would be the forecasts.  They also use a NB likelihood in the mentioned paper, so you're notebook has already been super helpful.
Thanks a lot for any inputs!

**Comment by murometz** [comment]
@tblazina Hello Tim, how have you adapted the loss to work with 3D input tensors of LSTM?
Thank you!

**Comment by tblazina** [comment]
@murometz, Unfortunately I haven't looked at this again in a long time so I'm not sure. I know the paper that I was attempting to implement was actually open sourced [here](https://github.com/awslabs/gluon-ts) some months ago. Good luck!
